### ISSUE TITLE
OIDC login fails for multiple users with different e-mail domains

### ISSUE THREAD

**Issue by kristianvld** [outline/outline#2678]
When you self host Outline and use the OIDC auth provider only people with the same domain in their email is able to log in.

Due to the restriction of only having one team when selfhosting, and you are automatically assigned a team based upon your email host name, two people with different domains in their emails are not able to log in to the same outline instance.

The first time you log in a team is created, if person had an @gmail.com address then other users that authenticate through the same OIDC provider will not be able to log in if they have anything else than @gmail.com, e.g. if they have an @outlook.com address etc.

The following console error is thrown:
```
ClientError [UnauthorizedError]: The maximum number of teams has been reached
    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)
    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:98:22) {
  redirectUrl: 'https://wiki.itemize.no',
  id: 'authentication_required'
}
```

I suspect that maybe this piece of code here is to blame. In my mind it would make more sense if we only looked up the team based upon the name, which is static when self-hosting, instead of looking up using the (name, domain, subdomain) tuple.
https://github.com/outline/outline/blob/51971d2c9af0ac3db6206e53b0a38062db308a81/server/routes/auth/providers/oidc.js#L93-L98

**Comment by almereyda** [comment]
If the actual error is here: `The maximum number of teams has been reached`, eventually we may also wish to discuss if the limitations of self-hosted Outline installations to one team keeps on being desireable.

It also appears open to debate for me, if the definition and equality of a Team with an email domain is always given in all circumstances, or if it makes things harder in the long run. Also Conway's Law might be at play here, that is do we structure our organisations after our tools, or do the tools allow us to model our digital environments after the actual structures of our groups and organisations?

**Comment by nikoPLP** [comment]
I managed to have any kind of email addresses be able to login, by using an OIDC (OpenID Connect) provider specially configured to hide the different domain names.
Namely, I installed a KeyCloak instance and configured it to be the OIDC provider of Outline.
In the keyCloak, users can register with their usual email address.
Once keyCloak passes the email address on to Outline, it hides the true email address of the user, and instead gives a virtual email address to Outline. This address is of the form: [username]@[domainNameOfOutline].
Let's say you have an outline instance running at `outline.mysite.com`, and you have a user that registered in KeyCloak with username `bob` and email address `famousbob@gmail.com`.
The email address that KeyCloak sends to Outline will be `bob@outline.mysite.com`.

How to configure keyCloak to achieve this outcome ?
* in `profile.properties` insert 2 lines
  * feature.scripts=enabled
  * feature.upload_scripts=enabled
* restart keycloak
* then in the admin console of keyCloak, enter your OIDC outline client configuration, in `mappers`, create a new `script mapper`.
*  for the `email` token claim, enter the following script
```
exports = user.getUsername() + "@outline.mysite.com"
```
* then if you want the emails sent by outline (notifications) to be received by your users, you will need to configure your MTA to have aliases that map every `*@outline.mysite.com" back to the original email address, in our case, the gmail.com address. The aliases should be configured to be an RCPT replacement for all emails that have an `*@outline.mysite.com` RCPT.
* in OpenSMTP, you do that with :
```
table team file:/etc/mail/team
action "forward" forward-only virtual <team> 
match from any for domain outline.mysite.com "forward"
```
and the `/etc/mail/team` is an aliases file of the usual form:
```
bob@outline.mysite.com: famousbob@gmail.com
```
don't forget to run `cd /etc/mail/ ; makemap team ; rcctl restart smtp`

you will have to adapt that to your MTA (postfix or else)

this way, all email addresses known to Outline are from the same domain, hence, from the same team.
